docs/proposals/project-user-resource-quotas.md

```diff
@@ -0,0 +1,209 @@
+The reasons vary from being able to stop the depletion of resources in a shared datacenter, to being able to sell packages (projects) with a set 
+number of resources available. 
+
+This will be an EE feature.
+
+## How will the resource quotas work
+
+Project quota - amount of resources (vCPU, RAM, storage) that can allocated for nodes in a project.
+
+Rules:
+
+- The quotas are for node resources
+- The quotas are enforced(we should block creation of new nodes that would exceed the quota)
+- The resource quotas management will be admin-only.
+- By default, projects will not have a quota, meaning they can use as many resources they want
+- There will be a way to set a default resource quota for all projects
+- When a node is deleted, the quota usage needs to decrease for the project
+- If the quota is exceeded(by lowering the quota under current capacity for example), it is not the responsibility of KKP to remove nodes. Admins should fix that in communication with the project/user.
+- The storage quota targets just the disk size of the node. It will only affect node-local PV storage.
+
+## Implementation
+
+### CRD Changes
+
+Add a new flexible ResourceQuota CRD which will hold the desired quota and current consumption of the quota.
+
+```go
+type ResourceQuota struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   ResourceQuotaSpec   `json:"spec,omitempty"`
+    Status ResourceQuotaStatus `json:"status,omitempty"`
+}
+
+type ResourceQuotaSpec struct {
+    // QuotaSubject describes the object for which the quota is applied to.
+    QuotaSubject *QuotaSubject `json:"quotaSubject"`
+    // ResourceQuotas is a map of maximum resource quotas per resource
+    ResourceQuotas corev1.ResourceList `json:"resourceQuotas"`
+}
+
+type QuotaSubject struct {
+    // Name of the quota subject
+    Name string `json:"name"`
+}
+
+type ResourceQuotaStatus struct {
+    // ResourceConsumption is map which holds the current usage of resources for all seeds
+    ResourceConsumption corev1.ResourceList `json:"resourceConsumption"`
+    // LocalConsumption is a map of the current usage of resources for the local seed
+    LocalConsumption corev1.ResourceList `json:"localConsumption"`
+
+}
+```
+
+The ResourceQuota is a master-cluster resource. It will need to be synced to seed clusters to allow webhooks access.
+As we have a distributed platform over multiple seeds, each seed will have to report its usage in the status. Then the master quota controller
+can recalculate the total resource consumption and re-sync the quotas on the seeds.
+
+### How to sync quota info from seed clusters and master cluster and back
+
+We would need a controller for ResourceQuotas in the master cluster: `master-resource-quota-controller` - MRQC
+
+The MRQC would watch master cluster ResourceQuotas. These are the ones that are created through the API, and maybe one day through `kubectl`.
```

Comment:
We already have a bunch of "...-sync" controllers that mirror master resources onto seed clusters. It might be possible/good to also separate this proposal into 2 controllers, once to implement the business logic (possibly running on the seed?), and one that just blindly mirrors things (running on the master cluster).

Just as a thought.

Reply:
I see it like this:
MRQC does the mirroring to and back from the seed clusters
A bit down is a paragraph about 2 more controllers:
1. One running as part of the `user-cluster-controller` which calculates the resource usage for the cluster and updates cluster status
2. One running as part of `seed-cluster-controller` which calculates the resource usage for a project on the Seed, and updates the ResourceQuotas `status.localConsumption`